app/domain/authentication/authn_azure/application_identity.rb

```diff
@@ -2,30 +2,45 @@ module Authentication
   module AuthnAzure
 
     Log = LogMessages::Authentication::AuthnAzure
+    Err = Errors::Authentication::AuthnAzure
 
     class ApplicationIdentity
 
-      def initialize(role_annotations:, service_id:, logger:)
+      # resources: list of AzureResource objects included in the Application Identity
+      # resource_types: list of the Azure resource types included in the Application Identity
+      attr_reader :resources, :resource_types
+
+      def initialize(role_annotations:, service_id:, azure_resource_types:, logger:)
         @role_annotations = role_annotations
         @service_id       = service_id
+        @azure_resource_types = azure_resource_types
         @logger           = logger
-      end
 
-      def constraints
-        @constraints ||= {
-          subscription_id:          constraint_value("subscription-id"),
-          resource_group:           constraint_value("resource-group"),
-          user_assigned_identity:   constraint_value("user-assigned-identity"),
-          system_assigned_identity: constraint_value("system-assigned-identity")
-        }.compact
+        init_resources
       end
 
       private
 
-      # check the `service-id` specific constraint first to be more granular
-      def constraint_value constraint_name
-        annotation_value("authn-azure/#{@service_id}/#{constraint_name}") ||
-          annotation_value("authn-azure/#{constraint_name}")
+      def init_resources
+        @resource_types = []
+        @resources = @azure_resource_types.each_with_object([]) do |resource_type, resources|
+          resource_value = resource_value(resource_type)
+          if resource_value
+            resources.push(
+              AzureResource.new(
+                resource_type: resource_type,
+                resource_value: resource_value
+              )
+            )
+            @resource_types.push(resource_type)
+          end
+        end
+      end
+
+      # check the `service-id` specific resource first to be more granular
+      def resource_value resource_type
+        annotation_value("authn-azure/#{@service_id}/#{resource_type}") ||
+          annotation_value("authn-azure/#{resource_type}")
```

Comment:
The method `annotation_value` below hits the database, making this no longer a value object.

Reply:
Also, I believe the way this is written and the how sequel works, this is making a separate database call to try every single possible valid annotation.  Which is both inefficient but also feels like a roundabout way to do it.  Instead, get all the annotations in a single batch inside the command class `ValidateApplicationIdentity` (where database calls are allowed), then pass them into a modified version of this class to get back a collection of `AzureResource` objects.

You'll then be inferring the resource type from the annotation itself, rather than trying all combinations of possible types and checking if an annotation exists for each of them.

Now, as for the collection of `AzureResource`s you return, you could make an custom collection object with a `resource_types` method similar to what you have now, but I don't think that's the simplest approach.  You could also just return an array of `AzureResource`s, and getting the `resource_types` for when you use them in:
```ruby
identifiers_constraints = application_identity.resource_types & identifiers
```
is just a simple `map(&:resource_type)` away.  

But I'd also strongly consider another approach.... will post separately.

Reply:
i am not sure that it calls the database each time because we initialize the `ApplicationIdentity` object like this:
```
def application_identity
        @application_identity ||= @application_identity_class.new(
          role_annotations: role.annotations,
          service_id:       @service_id,
          azure_resource_types: AZURE_RESOURCE_TYPES,
          logger:           @logger
        )
      end
```

then, inside `ApplicationIdentity` each call to `@role_annotations` will just call the given value to `role_annotations`. I may be wrong of course.

Anyway, i changed the calling method to:
```
      def application_identity
        @application_identity ||= @application_identity_class.new(
          role_annotations: role_annotations,
          service_id:       @service_id,
          azure_resource_types: AZURE_RESOURCE_TYPES,
          logger:           @logger
        )
      end

      def role_annotations
        @role_annotations ||= role.annotations
      end
```

so it fixes the issue and the method `annotation_value` does not hit the DB.